FSE Fix copy page flow. (WIP)

noahtallen opened the pull request:
Wow, this is a surprisingly difficult problem to solve because both Jetpack's copy_post and FSE are really hacking the existing filters/actions to load post content as they want and the approaches are not compatible.

**Problems fixed by this PR:**
1. The copy page function completely deletes the block template from the editor settings, so we now set the block template after that happens (priority 11).
2. On new pages, the existing post content is used as the "initial edits" to load into Gutenberg. That post content did not include the header/footer. We now merge the page template into the default post content for the editor instance.

However, though these are all necessary, the result of this PR is that the post content does not display when the post is copied, just the header/footer and post title. The good news is that we know what causes this bug. The bad news is that it's a systematic issue in Gutenberg: nested templates are f****d. (see [here](https://github.com/Automattic/wp-calypso/issues/35605#issuecomment-542931228) for more explanation)

We tell Gutenberg that our template is "template part, then post content, then another template part", but Gutenberg validates that against post content _including_ all of the nested post content blocks. There is no way to skip validation of nested blocks, which means when the editor is set up, those nested blocks are removed because they don't exist in the block template.

A solution might be to add all of the post content blocks as innerBlocks of the post content block in the block template, so that the blocks all exist in the template during validation. This is 💯hacky, but if we want this fixed for FSE soon, that's the way to go. I tried to get this working today, but couldn't get all the parsing working as expected, so it's not in this PR yet.

However, I already have https://github.com/WordPress/gutenberg/issues/11681 on my radar, and my plan is to work on a solution at the source as part of the work involved with that. 

### Testing instructions
Copy a page and see if the editor loads as expected :)
- SPT should not display
- Headers and footers should display.
- Post content from copied post should display. (note: not working)
- Additionally, smoke test normal page creation and editing routes.

Fixes #35605

noahtallen committed:
Remove check per Marko's suggestion

noahtallen committed:
Add header/footer data to default post content

noahtallen committed:
Remove log and fix typo

noahtallen committed:
Fix PHP error

kwight commented:
Great sleuthing @noahtallen  <3

Yes, Copy Post rips out any template to avoid templates that Post Formats might try to introduce, but adding FSE templates at a later priority is ✨ .

> If anyone would like to pick this up while I'm gone, the solution would be to add all of the post content blocks as innerBlocks of the post content block in the block template, so that the blocks all exist in the template during validation. This is 💯hacky, but if we want this fixed for FSE soon, that's the way to go.

Since https://github.com/WordPress/gutenberg/issues/11681 is a little up in the air, could this be split into two PRs?

- just the `load_starter_page_templates` change to close https://github.com/Automattic/wp-calypso/issues/35604 quickly, and
- the rest with the `dispatch( 'core/block-editor' ).setTemplateValidity( true );` hack until https://github.com/WordPress/gutenberg/issues/11681 has a solution?

noahtallen commented:
Note: I moved the SPT part of this to #36874. (Still leaving the code here because it's easier to troubleshoot with that bug also fixed. It'll "disappear" once we rebase this branch on top of master after that PR is merged.)

noahtallen commented:
> the rest with the dispatch( 'core/block-editor' ).setTemplateValidity( true ); hack until WordPress/gutenberg#11681 has a solution?

I think we _could_ merge the rest of this code as is, but I'd recommend against it. Though it fixes theoretical holes in the flow, the existing bug just doesn't show the header/footer on the copied page, but they will show up if the user saves the page or publishes it. With this partial fix, none of the page content would get copied at all, and I think if the user were to publish the "copied" page, it would be blank since Gutenberg overrides the post content to match the FSE template.

Note that the template validity hack doesn't fix this bug because Gutenberg is manually synchronizing the initial edits with the template. A fix for that bug should fix this too as long as behavior in template validation and template synchronization is the same.

vindl commented:
> I think we could merge the rest of this code as is, but I'd recommend against it. Though it fixes theoretical holes in the flow, the existing bug just doesn't show the header/footer on the copied page, but they will show up if the user saves the page or publishes it.

@noahtallen I agree that we should probably not merge this as is. What do you think about closing this for now until the required Core work is settled?

noahtallen commented on apps/full-site-editing/full-site-editing-plugin/full-site-editing/class-full-site-editing.php:
```diff
@@ -516,7 +549,7 @@ private function fse_map_block_to_editor_template_setting( $block ) {
 
 		$inner_blocks_template = array();
 		foreach ( $inner_blocks as $inner_block ) {
-			$inner_blocks[] = fse_map_block_to_editor_template_setting( $inner_block );
```
I'm not sure! I was troubleshooting this part of code and realized the mistake :)

noahtallen commented on apps/full-site-editing/full-site-editing-plugin/full-site-editing/class-full-site-editing.php:
```diff
@@ -615,11 +648,6 @@ public function restrict_template_taxonomy_deletion( $term, $taxonomy ) { // php
 		global $submenu;
 
-		// For safety.
-		if ( ! \A8C\FSE\is_full_site_editing_active() ) {
```
yep!

noahtallen commented:
> What do you think about closing this for now 

IMO we could just leave it on hold and then rebase once core work is ready. If you think it would be better to open a new PR when it's ready to go, that also works for me.

noahtallen commented:
Note: re-check that everything works after https://github.com/Automattic/wp-calypso/pull/37500 is merged. (Probably would be a good idea to integrate the approaches a bit more per disabling SPT on the copy page flow.)
